[local_auth_darwin] macOS Support

alexrabin-sentracam opened the pull request:
Adds macOS support for local_auth_darwin

![Screenshot 2024-03-05 at 8 30 35 AM](https://github.com/flutter/packages/assets/160153899/89bcfa78-b998-401e-869c-28b9d82a9229)

![Screenshot 2024-03-05 at 8 30 56 AM](https://github.com/flutter/packages/assets/160153899/69f0e215-1a7c-45eb-99a6-264458b0e771)

## Cancelled Example:

![Screenshot 2024-03-05 at 8 31 12 AM](https://github.com/flutter/packages/assets/160153899/1196b4e9-c010-4e96-994b-7467d1561ad1)

## Success Example

![Screenshot 2024-03-05 at 8 31 32 AM](https://github.com/flutter/packages/assets/160153899/acd0d550-3be2-46cf-957c-fbbe445abfa4)

## Error Example
<img width="912" alt="Screenshot 2024-03-05 at 4 01 58 PM" src="https://github.com/flutter/packages/assets/160153899/3a16eed5-d8b1-42a2-b6ab-ca82ade101ce">


*List which issues are fixed by this PR. You must list at least one issue.*

https://github.com/flutter/flutter/issues/140685

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I [linked to at least one issue that this PR fixes] in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

alexrabin-sentracam commented:
@stuartmorgan That pr doesn't seem to have any momentum and is outdated so I figured I would take this on.

stuartmorgan commented on packages/local_auth/local_auth_darwin/CHANGELOG.md:
```diff
@@ -1,12 +1,16 @@
 ## 1.2.2
 
-* Adds compatibility with `intl` 0.19.0.
+- Adds compatibility with `intl` 0.19.0.
```
Please do not reformat the file.

alexrabin-sentracam replied:
Oops, my prettier on my vs code must have done that automatically. Will fix

stuartmorgan commented on packages/local_auth/local_auth_darwin/CHANGELOG.md:
```diff
@@ -1,12 +1,16 @@
+## 2.0.0
+
+- Adds MacOS Support. Requires MacOS 10.15 or higher
```
macOS

stuartmorgan replied:
"Support" is not a title or proper noun, so should not be capitalized.

stuartmorgan commented on packages/local_auth/local_auth_darwin/CHANGELOG.md:
```diff
@@ -1,12 +1,16 @@
+## 2.0.0
+
+- Adds MacOS Support. Requires MacOS 10.15 or higher
```
Where does the 10.15 requirement come from?

alexrabin-sentracam replied:
Since we are using LABiometryTypeFaceID that requires macOS 10.15 or higher https://developer.apple.com/documentation/localauthentication/labiometrytype/labiometrytypefaceid?language=objc

stuartmorgan replied:
Why not just not support face ID when running on macOS 10.14?

alexrabin-sentracam replied:
The LABiometryTypeFaceID enum value is only available on macOS 10.15 or higher. We can't use this value on macOS 10.14

alexrabin-sentracam replied:
Fixed

stuartmorgan commented on packages/local_auth/local_auth_darwin/README.md:
```diff
@@ -1,6 +1,6 @@
 # local_auth_darwin
 
-The iOS implementation of [`local_auth`][1].
+The iOS and MacOS implementation of [`local_auth`][1].
```
macOS

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -153,6 +153,26 @@ - (void)showAlertWithMessage:(NSString *)message
+  [alert beginSheetModalForWindow:NSApp.keyWindow
+                completionHandler:^(NSModalResponse returnCode) {
+                  if (returnCode == NSAlertSecondButtonReturn) {
+                    NSURL *url = [NSURL URLWithString:@"x-apple.systempreferences:com.apple."
```
https://github.com/flutter/packages/pull/5766/files#r1486380427

alexrabin-sentracam replied:
https://gist.github.com/rmcdongit/f66ff91e0dad78d4d6346a75ded4b751

stuartmorgan replied:
You've marked this as resolved, but I don't see a code comment in the current version, as requested by the linked review.

More importantly though, what the link you provided appears to be describing is a reverse-engineered system that uses private keys and has a proven track record of fragility to OS updates. That means that:
- there's no guarantee that Apple would consider this app-store compliant, and
- it is very likely to break without warning in the future, leaving a button that doesn't actually work.

Am I misunderstanding the discussion there? If not, then this needs a different solution for macOS that is more robust even if it's not as featureful.

alexrabin-sentracam replied:
What do you suggest then in order to open the settings app? Here is another source with all the url schemes for settings: https://www.mbsplugins.de/archive/2020-04-05/MacOS_System_Preference_Links

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -153,6 +153,26 @@ - (void)showAlertWithMessage:(NSString *)message
+  if (openSettingsButtonTitle != nil) {
+    [alert addButtonWithTitle:openSettingsButtonTitle];
+  }
+  [alert beginSheetModalForWindow:NSApp.keyWindow
```
See https://github.com/flutter/packages/pull/5766/files#r1486389711 for discussion of the window that should be used; there is no guarantee that the Flutter content is in the key window.

alexrabin-sentracam replied:
I am not seeing how to get the current window with the registrar. Can you link me to some documentation please?

alexrabin-sentracam replied:
Nvm I found it

alexrabin-sentracam replied:
NSWindow* keyWindow = self.registrar.view.window;

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -257,12 +279,14 @@ - (void)handleError:(NSError *)authError
 
 #pragma mark - AppDelegate
 
+#if TARGET_OS_IOS
```
This should have a comment explaining why it's not relevant for macOS.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/local_auth_darwin.podspec:
```diff
@@ -16,8 +16,10 @@ Downloaded by pub (not CocoaPods).
-  s.dependency 'Flutter'
+  s.ios.dependency 'Flutter'
+  s.osx.dependency 'FlutterMacOS'
   s.platform = :ios, '12.0'
```
This should be updated to use the same format as the new line below it.

alexrabin-sentracam replied:
What do you mean?

stuartmorgan replied:
`s.ios.deployment_target = ...`

stuartmorgan commented on packages/local_auth/local_auth_darwin/example/ios/Runner.xcodeproj/project.pbxproj:
```diff
@@ -227,7 +227,7 @@
 			isa = PBXProject;
 			attributes = {
-				LastUpgradeCheck = 1510;
+				LastUpgradeCheck = 1430;
```
Please don't downgrade the iOS project metadata.

stuartmorgan commented on packages/local_auth/local_auth_darwin/example/ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Scheme
-   LastUpgradeVersion = "1510"
+   LastUpgradeVersion = "1430"
```
Same.

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/types/auth_messages_darwin.dart:
```diff
@@ -6,12 +6,12 @@ import 'package:flutter/foundation.dart';
 /// Provides default values for all messages.
 @immutable
-class IOSAuthMessages extends AuthMessages {
+class DarwinAuthMessages extends AuthMessages {
```
See https://github.com/flutter/packages/pull/5766#issuecomment-1878641342 and https://github.com/flutter/packages/pull/5766#discussion_r1486398912

stuartmorgan replied:
(Also, renaming public classes is a breaking change, which this doesn't otherwise need to be.)

alexrabin-sentracam replied:
I put back iOSAuthMessages and created MacOSAuthMessages

stuartmorgan commented on packages/local_auth/local_auth_darwin/example/macos/RunnerTests/RunnerTests.swift:
This file should not exist; macOS should be running the same native unit tests as iOS is, sharing the file in `darwin/Tests/`.

stuartmorgan commented on packages/local_auth/local_auth_darwin/pubspec.yaml:
```diff
@@ -2,7 +2,7 @@ name: local_auth_darwin
 repository: https://github.com/flutter/packages/tree/main/packages/local_auth/local_auth_darwin
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+local_auth%22
-version: 1.2.2
+version: 2.0.0
```
Once my comments about strings are resolved, this should be able to be 1.3.0.

alexrabin-sentracam commented:
@stuartmorgan  I have addressed all your comments

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -130,12 +132,31 @@ - (nullable NSNumber *)deviceCanSupportBiometricsWithError:
+      }
+#endif
+
+#if TARGET_OS_IOS
```
Why is all of this code duplicated and ifdef'd rather than the `@available` check including iOS?

alexrabin-sentracam replied:
When I try to combine both availability statements I get this error:
```
 @available does not guard availability here; use if (@available) instead [-Wunsupported-availability-guard]
       if (@available(macOS 10.15, *) || @available(iOS 11.0, *)) {
```

That was why I put the comment saying @available is not a BOOl expression and is why I had to separate the if statements. What would you suggest we do then?

alexrabin-sentracam replied:
Nvm fixed: https://forums.developer.apple.com/forums/thread/5237

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -130,12 +132,31 @@ - (nullable NSNumber *)deviceCanSupportBiometricsWithError:
                            error:&authError]) {
     if (authError == nil) {
+#if TARGET_OS_OSX
+      // @available(macOS 10.15, *) is not a BOOL expression.
```
This comment doesn't seem necessary; this is standard language behavior.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -49,13 +49,15 @@ - (instancetype)initWithOptions:(nonnull FLADAuthOptions *)options
 + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
   FLALocalAuthPlugin *instance = [[FLALocalAuthPlugin alloc] init];
   [registrar addApplicationDelegate:instance];
+  instance.registrar = registrar;
```
This should be an init parameter, as it's both required and unchanging.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -153,6 +174,27 @@ - (void)showAlertWithMessage:(NSString *)message
+  if (openSettingsButtonTitle != nil) {
+    [alert addButtonWithTitle:openSettingsButtonTitle];
+  }
+  NSWindow *keyWindow = self.registrar.view.window;
```
"Key window" is a specific term with a defined meaning in macOS development; this is not the key window. It's the window containing the Flutter content.

There's no ambiguity here though, so the variable can just be called `window`.

alexrabin-sentracam replied:
Fixed

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/messages.g.m:
```diff
@@ -41,7 +41,7 @@ - (instancetype)initWithValue:(FLADAuthResult)value {
 @end
 
-/// Pigeon equivalent of the subset of BiometricType used by iOS.
+/// Pigeon equivalent of the subset of BiometricType used by iOS & macOS.
```
Please see line 4 of this file; you are hand-editing a generated file rather than the source of generation.

alexrabin-sentracam replied:
Reverted

stuartmorgan reviewed (changes requested):
Based on the CI failure message I suspect you didn't add any source files to the RunnerTest target's build.

stuartmorgan commented:
> > Based on the CI failure message I suspect you didn't add any source files to the RunnerTest target's build.
> 
> I had to remove them since the CI said there were duplicate tests from the iOS project

This won't pass CI without native unit tests. You'll need to provide an exact error message you were getting from having them, since that description doesn't make sense to me. We have this pattern in a number of other plugins.

alexrabin-sentracam commented:
> This won't pass CI without native unit tests. You'll need to provide an exact error message you were getting from having them, since that description doesn't make sense to me. We have this pattern in a number of other plugins.

Okay I'll put back the tests then

alexrabin-sentracam commented:
@stuartmorgan I found this reference for all the url schemes that Apple has within the settings app: https://www.mbsplugins.de/archive/2020-04-05/MacOS_System_Preference_Links. If theses are up to par with you then what do you suggest we do for the settings button?

stuartmorgan commented:
> @stuartmorgan I found this reference for all the url schemes that Apple has within the settings app: https://www.mbsplugins.de/archive/2020-04-05/MacOS_System_Preference_Links.

That's not official documentation from Apple, so it doesn't address my earlier comments.

> If theses aren't up to par with you then what do you suggest we do for the settings button?

If you are not able to find a supported way to open settings programmatically, then I wouldn't attempt to provide a button that claims that it will do that, and would instead provide a high-level explanation of what the user needs to do.

stuartmorgan commented:
It still looks from the error message like the RunnerTests target is empty. Are you sure the file is added to that target? Are you able to run those tests in Xcode locally?

alexrabin-sentracam commented:
@stuartmorgan I found an article talking about how to find all the available app link URL's that Apple Supports and the plist keys that are needed to support them: https://www.linkedin.com/pulse/creating-web-links-launch-mac-apps-preferences-script-nick-tong

`sudo grep -A1 -r "NSPrefPaneAllowsXAppleSystemPreferencesURLScheme" /System/Library/PreferencePanes`

```
<key>NSPrefPaneAllowsXAppleSystemPreferencesURLScheme</key>
<true/>
```

I updated the Readme accordingly.

I also added the Tests for macOS.

stuartmorgan commented:
> @stuartmorgan I found an article talking about how to find all the available app link URL's that Apple Supports and the plist keys that are needed to support them: https://www.linkedin.com/pulse/creating-web-links-launch-mac-apps-preferences-script-nick-tong

Please read my comments above; a random person posting an article on LinkedIn is not the same as official documentation from Apple.

> I updated the Readme accordingly.

Adding an undocumented key that is used *for preference panes* to an application is unlikely to do anything, so is not something we would instruct people to do.

> I also added the Tests for macOS.

Based on the CI error, your change to the project appears to have broken the ability to build the example app.

alexrabin-sentracam commented:
> > @stuartmorgan I found an article talking about how to find all the available app link URL's that Apple Supports and the plist keys that are needed to support them: https://www.linkedin.com/pulse/creating-web-links-launch-mac-apps-preferences-script-nick-tong
> 
> Please read my comments above; a random person posting an article on LinkedIn is not the same as official documentation from Apple.
> 
> > I updated the Readme accordingly.
> 
> Adding an undocumented key that is used _for preference panes_ to an application is unlikely to do anything, so is not something we would instruct people to do.

Okay I am going to remove the settings button in the pop up and just update the messaging to tell the user to go to the settings app manually.

alexrabin-sentracam commented:
@stuartmorgan @LouiseHsu is there anything else needed before this pr can be merged?

stuartmorgan commented:
Please see https://github.com/flutter/flutter/wiki/Tree-hygiene#tldr for expected review times. It's been less than a full business day since you posted the last version.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -71,6 +72,14 @@ - (instancetype)initWithContextFactory:(NSObject<FLADAuthContextFactory> *)facto
   return self;
 }
 
+- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
```
There should not be any codepaths that don't set the registrar. This class should follow standard Obj-C convention of having an init method that has every possible parameter, and any necessary convenience inits should delegate to that one.

I would not expect the `initWithContextFactory:` to be necessary any more, since tests should always be specifying all arguments, so you should be able to just add a parameter to that method.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -153,6 +166,19 @@ - (void)showAlertWithMessage:(NSString *)message
+  return;
+#endif
+
+#if TARGET_OS_IOS
```
These are mutually exclusive blocks, so should use an `#if/#else` construction, not two separate `#if`s.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
There's no testing of the alert construction, or that it's using the registry. There should be a test that validates that the alert uses the window of the view provided by the registry, and that the expected strings are used in the alert. That will mean adding an injectable alert factory, similar to the existing auth factory.

alexrabin-sentracam replied:
Added alert construction test

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/local_auth_darwin.dart:
```diff
@@ -2,19 +2,21 @@
 export 'package:local_auth_platform_interface/types/biometric_type.dart';
 
-/// The implementation of [LocalAuthPlatform] for iOS.
+/// The implementation of [LocalAuthPlatform] for iOS & macOS.
```
Please use "and" instead of "&", to match our standard style.

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/local_auth_darwin.dart:
```diff
@@ -93,11 +97,11 @@ class LocalAuthDarwin extends LocalAuthPlatform {
   Future<bool> isDeviceSupported() async => _api.isDeviceSupported();
 
-  /// Always returns false as this method is not supported on iOS.
+  /// Always returns false as this method is not supported on iOS & macOS.
```
"iOS or macOS"

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/local_auth_darwin.dart:
```diff
@@ -118,4 +122,26 @@ class LocalAuthDarwin extends LocalAuthPlatform {
+      goToSettingsButton: messages?.goToSettingsButton ?? goToSettings,
+      goToSettingsDescription:
+          messages?.goToSettingsDescription ?? macOSGoToSettingsDescription,
+      // TODO(stuartmorgan): The default's name is confusing here for legacy
```
What legacy reasons require a new constant to have a confusing name?

alexrabin-sentracam replied:
That was a comment that was copied from auth_messages_ios

stuartmorgan replied:
I'm aware of that. I wrote that comment when writing new code that used an existing constant that we couldn't change without breaking clients.

That context doesn't apply to a completely new constant, so I don't understand why the TODO and the confusing name were copied into this file.

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/src/messages.g.dart:
```diff
@@ -36,15 +36,16 @@ enum AuthResult {
 }
 
-/// Pigeon equivalent of the subset of BiometricType used by iOS.
+/// Pigeon equivalent of the subset of BiometricType used by iOS & MacOS.
```
"and"

Same throughout this file (and the rest of the PR if there are other instances).

stuartmorgan replied:
macOS

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/src/messages.g.dart:
```diff
@@ -36,15 +36,16 @@ enum AuthResult {
-/// Pigeon equivalent of the subset of BiometricType used by iOS.
+/// Pigeon equivalent of the subset of BiometricType used by iOS & MacOS.
 enum AuthBiometric {
+  /// Face is only used on iOS
```
This is missing a period.

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/types/auth_messages_macos.dart:
```diff
@@ -0,0 +1,107 @@
+  /// Constructs a new instance.
+  const MacOSAuthMessages({
+    this.lockOut,
+    this.goToSettingsButton,
```
This is no longer used.

stuartmorgan replied:
Sending a string from Dart to the host isn't using the string in a meaningful sense, it's just passing it around. The reason these strings exist is to provide them to native code in order to display them to the user in a native dialog. There is no longer any code path that results in this string being displayed to the user, so it's dead code.

alexrabin-sentracam replied:
If I remove MacOSAuthMessages, should I rename iOSAuthMessages to DarwinAuthMessages?

stuartmorgan replied:
> If I remove MacOSAuthMessages

You should not remove `MacOSAuthMessages`. You should remove the parts of it that are never used.

> should I rename iOSAuthMessages to DarwinAuthMessages?

No; as discussed [here](https://github.com/flutter/packages/pull/6267#discussion_r1546793569) they are not the same. That would also re-create a different version of the same problem: a public API surface that allows setting some values for macOS that will never be used on macOS.

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/types/auth_messages_macos.dart:
```diff
@@ -0,0 +1,107 @@
+/// It shows in a dialog on macOS.
+String get macOSLockOut => Intl.message(
+    'Biometric authentication is disabled. Please lock and unlock your screen to '
+    'enable it.',
```
Are these instructions correct for macOS?

stuartmorgan commented on packages/local_auth/local_auth_darwin/lib/local_auth_darwin.dart:
```diff
@@ -40,7 +42,9 @@ class LocalAuthDarwin extends LocalAuthPlatform {
             sticky: options.stickyAuth,
             useErrorDialogs: options.useErrorDialogs),
-        _pigeonStringsFromAuthMessages(localizedReason, authMessages));
+        defaultTargetPlatform == TargetPlatform.macOS
```
This should be using `Platform.is*` (or a local-test-overridable wrapper around that), since it's interfacing with native code; if non-test code is executing this code with the default target overridden, it still needs to use the strings that correspond to the actual native code that will execute.

jmagman commented:
@LouiseHsu could you take a look at this one?

stuartmorgan commented:
Is this ready for re-review? There are still unaddressed comments and failing tests, so I assumed it was a work-in-progress.

alexrabin-sentracam commented:
I could use some assistance on the comments @stuartmorgan left me. I am not seeing a clean way to use the Platform.isMacos and still have the dart test cases pass. 

I am also not well versed in testing the alert construction as you requested.

stuartmorgan commented:
> I am not seeing a clean way to use the Platform.isMacos and still have the dart test cases pass.

Per my earlier comment, you can wrap those calls in something that can have an alternate implementation injected for unit tests, using standard dependency injection patterns.

alexrabin-sentracam commented:
> > I am not seeing a clean way to use the Platform.isMacos and still have the dart test cases pass.
> 
> Per my earlier comment, you can wrap those calls in something that can have an alternate implementation injected for unit tests, using standard dependency injection patterns.

All fixed. I addressed all your comments @stuartmorgan . This pr is ready for re-review.

stuartmorgan commented:
@alexrabin-sentracam I'll re-review when I have time (I currently have some very high priority work that will delay it); it's not necessary to constantly merge the branch in the meantime.

alexrabin-sentracam commented:
@stuartmorgan Okay sounds good. Thank you.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin.m:
```diff
@@ -257,12 +302,15 @@ - (void)handleError:(NSError *)authError
 
 #pragma mark - AppDelegate
 
+// This method is called when the app is resumed from the background only on iOS
```
This is missing a period.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin_Test.h:
```diff
@@ -2,16 +2,33 @@
 - (LAContext *)createAuthContext;
 @end
 
+@protocol FLADAlertFactory <NSObject>
```
This needs a declaration comment, per the style guide.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin_Test.h:
```diff
@@ -2,16 +2,33 @@
+@protocol FLADAlertFactory <NSObject>
+
+#if TARGET_OS_OSX
+- (NSAlert *)createNSAlert;
```
As do the methods.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Classes/FLALocalAuthPlugin_Test.h:
```diff
@@ -2,16 +2,33 @@
+@end
+
 @interface FLALocalAuthPlugin ()
 /// Returns an instance that uses the given factory to create LAContexts.
```
This comment needs to be updated to reflect the changes.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -49,11 +92,30 @@ - (void)setUp {
   self.continueAfterFailure = NO;
 }
 
+- (StubAlertFactory *)createStubAlertFactory {
```
This needs a declaration comment, per style guide.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -38,6 +38,49 @@ - (LAContext *)createAuthContext {
 
+@interface StubAlertFactory : NSObject <FLADAlertFactory>
+#if TARGET_OS_OSX
+@property(copy, nonatomic) NSAlert *alert;
```
Why are these properties `copy`?

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -38,6 +38,49 @@ - (LAContext *)createAuthContext {
+- (instancetype)initWithNSAlert:(NSAlert *)alert {
+  self = [super init];
+  if (self) {
+    _alert = alert;
```
These aren't doing copies, which violates the current API contract. (This will be resolved if you don't declare them as copy though.)

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -166,12 +237,70 @@ - (void)testFailedAuthWithBiometrics {
+                         // Tests that the alert window is the same as the registrar window
+                         XCTAssertEqual(alertFactory.alert.window, reg.view.window);
+                         // TODO(stuartmorgan): Fix this; this was the pre-Pigeon-migration
+                         // behavior, so is preserved as part of the migration, but a failed
```
How can an entirely new platform implementation have previous behavior that needs to be preserved?

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -49,11 +92,30 @@ - (void)setUp {
+  StubAlertFactory *alertFactory;
+#if TARGET_OS_OSX
+
+  id mockNSAlert = OCMClassMock([NSAlert class]);
```
Why is this a mock? The only thing that this appears to be used for in the test is to read a public property back out of it, which can be done with a real object, avoiding the need for a mock.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -166,12 +237,70 @@ - (void)testFailedAuthWithBiometrics {
+                                    FlutterError *_Nullable error) {
+                         XCTAssertTrue([NSThread isMainThread]);
+                         // Tests that the alert window is the same as the registrar window
+                         XCTAssertEqual(alertFactory.alert.window, reg.view.window);
```
I would expect there to also be a test that if the user provides a custom message, the alert has that message.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.m:
```diff
@@ -166,12 +237,70 @@ - (void)testFailedAuthWithBiometrics {
+  [self waitForExpectationsWithTimeout:kTimeout handler:nil];
+}
+
+#endif
```
You've changed the iOS codepaths to make the alert testable; there needs to be at least one new test covering those changes.

jmagman commented:
@alexrabin-sentracam looks like there's some merge conflicts, let us know when this is ready to re-review!

jmagman commented:
@LouiseHsu would you mind taking a look at this?

stuartmorgan commented:
Is this actually ready for re-review? There are still open comments from the last review, so I was assuming it was not yet ready.

alexrabin-sentracam commented:
@stuartmorgan I am seeing this error in the LUCI after the tests were converted to swift. 

> xcodebuild: error: Failed to build workspace Runner with scheme Runner.: Cannot test target “RunnerTests” on “My Mac”: My Mac’s macOS 13.6.1 doesn’t match RunnerTests’s macOS 14.0 deployment target.
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8746482769422894929/+/u/Run_package_tests/native_test/stdout?format=raw

Do you have any recommendations for what I should do?

stuartmorgan commented:
> @stuartmorgan I am seeing this error in the LUCI after the tests were converted to swift.

Why do you believe the error is related to the tests being converted to Swift? You changed the `MACOSX_DEPLOYMENT_TARGET` in [this commit](https://github.com/flutter/packages/pull/6267/commits/a91e4e912d9d7e33d3dc33a38600130481acdb76#diff-1464ddfcf50b83c234b3529d7d64d867e68def0e1e0b95c9a5ff3944f7aa1a37).

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/OCSwiftMock.h:
Please do not re-introduce OCMock; I removed it intentionally. New mocks/stubs should use the same pattern the other tests now use, which is to introduce a protocol that has the desired methods, and then have tests implement that protocol. See [here](https://github.com/flutter/packages/blob/decb46042bc5b000c0de27f2ff1ab4beeaa7b921/packages/local_auth/local_auth_darwin/darwin/local_auth_darwin/Sources/local_auth_darwin/include/local_auth_darwin/FLALocalAuthPlugin_Test.h#L10-L20), [here](https://github.com/flutter/packages/blob/main/packages/local_auth/local_auth_darwin/darwin/local_auth_darwin/Sources/local_auth_darwin/FLALocalAuthPlugin.m#L11-L56), and [here](https://github.com/flutter/packages/blob/decb46042bc5b000c0de27f2ff1ab4beeaa7b921/packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.swift#L26-L70) for an example of the pattern. (I would expect the new wrappers you would need to be simpler than that one.)

alexrabin-sentracam commented:
> @alexrabin-sentracam do you need some help with the OCMock bit?

@LouiseHsu Yes please, I'd appreciate that

stuartmorgan commented:
> > @alexrabin-sentracam do you need some help with the OCMock bit?
> 
> @LouiseHsu Yes please, I'd appreciate that

I've pushed changes that restructure to avoid the need for OCMock. @alexrabin-sentracam some follow-up work is still needed:
- The changes to actually add OCMock to the project need to be undone; I didn't do any project changes.
- I did not have to change any tests as part of the work, which suggests that none of the tests are actually checking anything about the interactions with `NSAlert`/`UIAlertController` to make sure they are as expected. That's an important part of the test coverage. The structure I set up will allow for querying that information, but I didn't actually add the tests.

stuartmorgan commented:
From triage: @alexrabin-sentracam Are you still planning on updating this now that the OCMock removal has been handled?

alexrabin-sentracam commented:
@stuartmorgan I've been assigned higher priorities at my job. I don't know when I'll be able to take a look at this 😕

stuartmorgan reviewed (approved):
I've done the remaining cleanup, and added basic testing of the dialog display for both macOS and iOS.

@LouiseHsu this is ready for final review.

stuartmorgan commented on packages/local_auth/local_auth_darwin/darwin/Tests/FLALocalAuthPluginTests.swift:
```diff
@@ -310,14 +324,53 @@ class FLALocalAuthPluginTests: XCTestCase {
+    ) { resultDetails, error in
+      #if os(macOS)
+        expectation.fulfill()
+      #else
```
Good point; reworked to be a TODO comment above the `#if`, and removed the `#else`.

stuartmorgan commented:
(That last failure was just the current macOS host flake we're having, unlike the previous ones.)
